### ISSUE TITLE
Potential error with current implementation of add_variables of NoisyOrModel

### ISSUE THREAD

**Issue by gokul-uf** [pgmpy/pgmpy#319]
Suppose I have an NoisyOrModel instance of maximum cardinality (say 3)

``` python
noisy = NoisyOrModel(['x1', 'x2', 'x3'], [2, 3, 2], [[0.6, 0.4],
....                                 [0.2, 0.4, 0.7], [0.1, 0.4]])
```

Internally we now have a 3x3 numpy array for inhibitor_probability
now,

``` python
noisy.add_variables(['x4'],[4],[0.1,0.2,0.3,0.4])
```

would throw an error because of 

``` python
self.inhibitor_probability = np.concatenate((self.inhibitor_probability, inhibitor_probability_uni))
```

I suggest that we keep track of the maximum cardinality to fix this

**Comment by gokul-uf** [comment]
@abinashpanda @ankurankan

**Comment by gokul-uf** [comment]
### More Info

 to solve it we can have the hack of maintaining the maximum cardinality of variables added so far.

But even this has a problem,
if one of the variables has a very large cardinality (say 100) and others have cardinalities 2,3,5 .etc,
most of the numpy array values would be 0, a waste of memory and not scalable

Here's a stack-overflow discussion thread on the same http://stackoverflow.com/questions/1053928/very-large-matrices-using-python-and-numpy

Turns out scipy does have support for sparse matrices, we could use them instead of numpy 2D arrays, this could be one possible solution
http://docs.scipy.org/doc/scipy/reference/sparse.html

**Comment by ankurankan** [comment]
@gokul-uf Why don't you go ahead and implement this and then let's see how this works? Sparse matrices would be a good option in this case.

**Comment by liquidmetal** [comment]
I tried this out and did not see the error mentioned here. Seems like it was fixed in PR #338 - (it includes a commit called `Fixed NoisyOrModel and its tests.`) @ankurankan could you please confirm?

**Comment by ankurankan** [comment]
@liquidmetal Yes this was fixed. Closing.
